### PR TITLE
 TCH-38 - add cache for get product

### PR DESCRIPTION
**Pull request by imwillim**

### CODE REVIEW

File: spec/services/get_product_service_spec.rb [hunk 1]
```diff
@@ -39,15 +47,34 @@
...
+    context 'when a product exists in cache' do
+      before do
+        allow(CacheManager).to receive(:fetch_value).and_return(tea.to_json)
+      end
+      it 'return cached product' do
+        service.call
+
+        expect(service.success?).to eq true
+        expect(service.result['id']).to eq tea.id
+        expect(service.result['id']).not_to eq milk_tea.id
```
Comment (bmhuyquoc104) [inline spec/services/get_product_service_spec.rb:62]: only 1 product, if the id is equal to tea, it is obvious not eq to milk_tea => unnecessary expect

File: spec/services/get_product_service_spec.rb [hunk 2]
```diff
@@ -39,15 +47,34 @@
...
+      end
 
-      it 'returns tea' do
+      it 'return product from database' do
         service.call
 
         expect(service.success?).to eq true
-        expect(service.result.id).to eq tea.id
-        expect(service.result.id).not_to eq milk_tea.id
+        expect(service.result['id']).to eq(tea.id)
```
Comment (bmhuyquoc104) [inline spec/services/get_product_service_spec.rb:76]: where the expect for the set redis behavior after it get data from db ?

File: spec/services/get_product_service_spec.rb [hunk 3]
```diff
@@ -39,15 +47,34 @@
...
+        expect(service.result['id']).not_to eq milk_tea.id
```
Comment (bmhuyquoc104) [inline spec/services/get_product_service_spec.rb:77]: same error like above

File: spec/services/get_product_service_spec.rb [hunk 4]
```diff
@@ -18,6 +18,10 @@
     context 'when product_id is nil' do
       let(:product_id) { nil }
 
+      before do
+        allow(CacheManager).to receive(:fetch_value).and_return(nil)
```
Comment (bmhuyquoc104) [inline spec/services/get_product_service_spec.rb:22]: this validate will works on both cache and db level, if test only for cache, what happens for db level, => don't need to specific the validation from cache level, this should be a logic that not depend on either cache or db level => remove

**Comment by imwillim** [comment]
After running unit test, cache does not store any key-value pair (get-single-product)

image
image

**Review by bmhuyquoc104 — APPROVED** [review]

**Pull request merged**